tests/hf/test_tapering.py

```diff
@@ -499,3 +456,156 @@ def test_transform_hamiltonian(symbols, geometry, generator, paulix_ops, paulix_
+    [
+        (
+            ["H", "H"],
+            np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.40104295]], requires_grad=False),
+            0,
+            [
+                qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(1)]),
+                qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(2)]),
+                qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(3)]),
+            ],
+            2,
+            [1, -1, -1],
+        ),
+        (
+            ["H", "H", "H"],
+            np.array(
+                [[-0.84586466, 0.0, 0.0], [0.84586466, 0.0, 0.0], [0.0, 1.46508057, 0.0]],
+                requires_grad=False,
+            ),
+            1,
+            [
+                qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(2) @ qml.PauliZ(4)]),
+                qml.Hamiltonian([1.0], [qml.PauliZ(1) @ qml.PauliZ(3) @ qml.PauliZ(5)]),
+            ],
+            3,
+            [1, -1],
+        ),
+    ],
+)
+def test_optimal_sector(symbols, geometry, charge, generators, num_electrons, result):
+    r"""Test that find_optimal_sector returns the correct result."""
+    mol = qml.hf.Molecule(symbols, geometry, charge)
+    hamiltonian = qml.hf.generate_hamiltonian(mol)()
+
+    perm = optimal_sector(hamiltonian, generators, num_electrons)
+
+    assert perm == result
+
+
+@pytest.mark.parametrize(
+    ("symbols", "geometry", "generators"),
+    [
+        (
+            ["H", "H"],
+            np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.40104295]], requires_grad=False),
+            [
+                qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(1)]),
+                qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(2)]),
+                qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(3)]),
+            ],
+        ),
+    ],
+)
+@pytest.mark.parametrize(
+    ("num_electrons", "msg_match"),
+    [
+        (0, f"The number of active electrons must be greater than zero"),
+        (5, f"Number of active orbitals cannot be smaller than number of active electrons"),
+    ],
+)
+def test_exceptions_optimal_sector(symbols, geometry, generators, num_electrons, msg_match):
+    r"""Test that find_optimal_sector returns the correct result."""
+    mol = qml.hf.Molecule(symbols, geometry)
+    hamiltonian = qml.hf.generate_hamiltonian(mol)()
+
+    with pytest.raises(ValueError, match=msg_match):
+        optimal_sector(hamiltonian, generators, num_electrons)
+
+
+@pytest.mark.parametrize(
+    ("generators", "paulix_ops", "paulix_sector", "num_electrons", "num_wires", "result"),
+    [
+        (
+            [
+                qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(1)]),
+                qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(2)]),
+                qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(3)]),
+            ],
+            [qml.PauliX(wires=[1]), qml.PauliX(wires=[2]), qml.PauliX(wires=[3])],
+            (1, -1, -1),
+            2,
+            4,
+            np.array([1]),
```

Comment:
I actually wonder, how we can be sure that these states are correct? Can we reproduce them with an external reference such as qiskit?

Reply:
This might actually be difficult since the generators we get and the ones that Qiskit's z2symmetry functionality give are different. Still, let me try doing this once.

Reply:
You can verify the accuracy of the state by computing the HF energy as `<state|H|state>` for the tapered state and the tapered Hamiltonian. The code might be something like this:
```
o = np.array([1, 0])
l = np.array([0, 1])
state = functools.reduce(lambda i, j: np.kron(i, j), [l, l, o, o])
print((scipy.sparse.coo_matrix(state) @ qml.utils.sparse_hamiltonian(H) @ scipy.sparse.coo_matrix(state).T).toarray())
```

We should do such tests for some molecules externally, but not sure if we need this as a unit test.

Reply:
Well, as guessed Qiskit gives different results due to the different qubit ordering they follow. However, as far as I could correlate the individual answers for different molecules: `H2`, `H3+`, `LiH`, `BeH2`, the tapered HF states look okay to me.

Reply:
We need to check that the HF energy and VQE energy are correctly reproduced for few molecules with the tapered state and Hamiltonian. If we verify that the energies are correct, then the PR can be safely merged.

Reply:
I have tested the HF energies for tapered and untapered cases for the above molecules. They came out to be equal and seemed correct to me in general. I have added an unittest for this as well.